subsys/bluetooth/audio/has_client.c

```diff
@@ -418,307 +615,317 @@ static uint8_t active_preset_notify_cb(struct bt_conn *conn,
-					 struct bt_gatt_discover_params *params)
+static int control_point_subscribe(struct bt_has_client *inst)
 {
-	struct has_inst *inst = CONTAINER_OF(params, struct has_inst, params.discover);
-	const struct bt_gatt_chrc *chrc;
 	int err;
 
-	LOG_DBG("conn %p attr %p params %p", (void *)conn, attr, params);
+	__ASSERT_NO_MSG(inst->params == NULL);
+	inst->params = &inst->control_point_sub;
 
-	if (!attr) {
-		LOG_INF("Control Point not found");
-		discover_complete(inst);
-		return BT_GATT_ITER_STOP;
-	}
+	if (is_subscribed(&inst->control_point_sub)) {
+		LOG_DBG("Subscribed already");
+		control_point_subscribe_cb(inst->conn, 0, &inst->control_point_sub);
 
-	chrc = attr->user_data;
+		return 0;
+	}
 
-	err = control_point_subscribe(inst, chrc->value_handle, chrc->properties);
-	if (err) {
-		LOG_ERR("Subscribe failed (err %d)", err);
+	inst->control_point_sub.notify = control_point_notify_cb;
+	inst->control_point_sub.subscribe = control_point_subscribe_cb;
 
-		/* Cleanup instance so that it can be reused */
-		inst_cleanup(inst);
+	err = bt_gatt_subscribe(inst->conn, inst->params);
+	if (err != 0) {
+		LOG_ERR("bt_gatt_subscribe err %d", err);
+		inst->control_point_sub.notify = NULL;
+		inst->params = NULL;
 
-		discover_failed(conn, err);
+		return err;
 	}
 
-	return BT_GATT_ITER_STOP;
+	return 0;
 }
 
-static int control_point_discover(struct has_inst *inst)
+static bool is_control_point_supported(const struct bt_has_client *inst)
 {
-	LOG_DBG("conn %p", (void *)inst->conn);
-
-	(void)memset(&inst->params.discover, 0, sizeof(inst->params.discover));
-
-	(void)memcpy(&inst->params.uuid, BT_UUID_HAS_PRESET_CONTROL_POINT,
-		     sizeof(inst->params.uuid));
-	inst->params.discover.uuid = &inst->params.uuid.uuid;
-	inst->params.discover.func = control_point_discover_cb;
-	inst->params.discover.start_handle = BT_ATT_FIRST_ATTRIBUTE_HANDLE;
-	inst->params.discover.end_handle = BT_ATT_LAST_ATTRIBUTE_HANDLE;
-	inst->params.discover.type = BT_GATT_DISCOVER_CHARACTERISTIC;
-
-	return bt_gatt_discover(inst->conn, &inst->params.discover);
+	return (inst->control_point_sub.value != 0 && inst->control_point_sub.ccc_handle != 0 &&
+		inst->active_index_sub.value != 0 && inst->active_index_sub.ccc_handle != 0);
 }
 
-static void features_update(struct has_inst *inst, const void *data, uint16_t len)
+static void features_changed(struct bt_has_client *inst, const void *data, uint16_t len)
 {
+	enum bt_has_features features;
 	struct net_buf_simple buf;
 
 	net_buf_simple_init_with_data(&buf, (void *)data, len);
 
-	inst->has.features = net_buf_simple_pull_u8(&buf);
+	features = net_buf_simple_pull_u8(&buf);
+
+	LOG_DBG("conn %p features 0x%02x", (void *)inst->conn, features);
+
+	if (features == inst->features) {
+		/* Not updated */
+		return;
+	}
+
+	inst->features = features;
+
+	/* Omit if not connected yet */
+	if (inst->state != STATE_CONNECTED) {
+		return;
+	}
```

Comment:
When can this happen?

Reply:
It happens when the notification is received before the application is notified about the service connection, i.e. the HAS has not been subscribed for all characteristics (so the discovery/connection procedure has not been completed yet).

Reply:
If we aren't subscribed, then why would we receive the notification? 

Isn't this already done in GATT in `bt_gatt_notification` where notifications for unsubscribed handles are ignored?